### PR TITLE
Android build fixes for c++11

### LINKED ISSUE #6750: Android: Irrlicht patch has wrong line endings
Since 2507d32afe05c73bb27ed221c2a592b0894fdc68 the included [irrlicht-native_activity.patch](https://github.com/minetest/minetest/commit/2507d32afe05c73bb27ed221c2a592b0894fdc68#diff-daa8d52a16bcc47ecc3d0db26d566e6c) has the wrong line endings and now fails to apply. I am not sure if I did something wrong here yet the same patch works fine when pulled from the original PR's feature branch of my own repo??

I'd make another PR myself, though I fear the same thing would happen again. Does anyone have any idea what could have happened here?

This is a diff of the diff, the latter part being the correct version (copied from console)
```
 @@ -8,6 +8,9 @@
- 
- #include "irrString.h"
- #include "os.h"
-+#if defined(_IRR_COMPILE_WITH_ANDROID_DEVICE_)
-+#include <android/native_activity.h>
-+#endif
- 
- namespace irr
- {
+ ^M
+ #include "irrString.h"^M
+ #include "os.h"^M
++#if defined(_IRR_COMPILE_WITH_ANDROID_DEVICE_)^M
++#include <android/native_activity.h>^M
++#endif^M
+ ^M
+ namespace irr^M
+ {^M
```

### PR DESCRIPTION
**Pull request by stujones11**
This PR fixes the android build for c++11 using clang as the default compiler. Hopefully this also fixes #6750 & #6588

Notes:

Now uses `c++_shared` STL (maybe someone else has better luck with static linking)
Currently only tested on armv7 although x86 builds successfully and is now reported to work.

Credit and thanks to @basicer for providing the engine patches.

### CODE REVIEW

**Comment by rubenwardy** [comment]
LGTM! Does this work?

**Comment by rubenwardy** [comment]
Code looks fine for me, but it's worth fixing the x86 build if possible

**Comment by ThomasMonroe314** [comment]
I use an x86 computer to play on mostly, so please fix that build :stuck_out_tongue:

**Comment by rubenwardy** [comment]
> I use an x86 computer to play on mostly, so please fix that build :stuck_out_tongue:

Only the x86 **Android** build doesn't work, from what I understand

**Comment by stujones11** [comment]
> Only the x86 Android build doesn't work, from what I understand

That is correct, I can confirm that the regular x86 linux build is unaffected by this, anyway.

**Comment by stujones11** [comment]
I spotted one small mistake, however, this does not seem to fix the problem.

**Comment by ThomasMonroe314** [comment]
oh :man_facepalming:  my bad, sorry

**Comment by stujones11** [comment]
My bad really, I can see how that could have been misinterpreted. I have since edited the top post to make it more clear.

**Comment by stujones11** [comment]
I have tried to fix the android-x86 build but the emulator itself is crashing before I can get any useful debug info. Since I can't do any more without access to proper hardware, the WIP label can be removed for my part.

I personally suspect openssl to be the problem here but have to evidence to support that. Logcat also shows that there may be issues with `MinetestAssetCopy.java` again, not something I am able to fix.

**Comment by Wayward1** [comment]
I have some interesting news: I went ahead and tried the x86 build again to see if I could get anything useful from the logcat, and suddenly it worked fine! The menu loaded, I started and played a singleplayer world just fine. I honestly can't explain it.

**Comment by stujones11** [comment]
> I honestly can't explain it. Same apk, same device.

I get this occasionally with my armv7 device, I always have done, even with older versions and forks. Also, it never has shutdown properly and will always crash when attempting to re-start from that state.

Again, these have always been issues for me, therefore beyond the scope of this PR.

**Comment by stujones11** [comment]
Squashed, rebased, re-fixed and re-tested though I am not sure how much longer I can maintain this :)

File: src/porting_android.h [hunk 1]
```diff
@@ -76,4 +76,9 @@ int getInputDialogState();
  */
 std::string getInputDialogValue();
 
+#ifndef SERVER
+float getDisplayDensity();
```
Comment (nerzhul) [inline src/porting_android.h]: i think those functions should be part of RenderingEngine class

**Comment by stujones11** [comment]
These are android specific functions, I can see so any easy way or reason to move these completely into rendering engine but perhaps I misunderstand. If you can explain exactly what you would like me to do I will happy to oblige.

**Comment by Wayward1** [comment]
The build currently fails for me with this output: https://pastebin.com/ziDGdc1S

**Comment by stujones11** [comment]
@Wayward1 I did have to force push some things after the game UI changes, are you sure you are using the most recent version?

**Comment by Wayward1** [comment]
Yes, I did a full re-clone and rebuild. I just re-cloned and rebuilt again, from this branch, with the same result: https://pastebin.com/1g0SwNgk

**Comment by stujones11** [comment]
Sorry, I should have spotted this yesterday but it was getting late for me.

This is the problem #6894

**Comment by Wayward1** [comment]
Yeah, I saw and tried that too, but it still fails for some reason (note: I just ran cmake in the existing directory). I'll try a fresh clone, run cmake, and build again once my build device is available

**Comment by stujones11** [comment]
It would probably be better to do as suggested in that issue and change the include line in `script/cpp_api/s_base.h`, however, running `cmake` as you would for a regular build generates the missing file and fixed the issue for me.

As this is a recently introduced bug, I do feel it the responsibility of this PR to fix it :P

**Comment by stujones11** [comment]
Rebased again @nerzhul could please reply to https://github.com/minetest/minetest/pull/6796#issuecomment-356368200

File: src/porting_android.h (hunk 1 above)
Comment (stujones11) [reply src/porting_android.h]: RenderingEngine already has wrapper functions for these https://github.com/minetest/minetest/pull/6796/files#diff-b6a7a4bb97008d7e6dfe76e614fbfb02R574

**Comment by Wayward1** [comment]
I'm not sure if it's an error on my part again, but the build now fails with https://pastebin.com/P7ens4yG

**Comment by stujones11** [comment]
I looks like the irrlicht patch didn't apply correctly, try just building irrlicht again and check your build log.

**Comment by Wayward1** [comment]
Ah, okay. Thank you, that fixed that issue, but now I keep getting errors with missing includes all over the place, like libint, luajit, leveldb and hiredis (until I disabled those two) even after running cmake. I'll try a fresh re-clone and rebuild.

**Comment by stujones11** [comment]
Yeah, that will likely be necessary. I just did that myself to be sure and it worked okay for me.

**Comment by Wayward1** [comment]
Oddly enough, I still get the same result:
``In file included from jni/../jni/src/map.cpp:51:
jni/src/database/database-leveldb.h:28:10: fatal error: 'leveldb/db.h' file not found
#include "leveldb/db.h"``

I disable that, and another one springs up

**Comment by stujones11** [comment]
I did not test using leveldb but will look at that asap. What error do you get with leveldb disabled?

**Comment by stujones11** [comment]
What I find most strange is that none of this has changed since you reported it working. Only some minor merge conflicts in `Android.mk` due the addition of new cpp files.

**Comment by Wayward1** [comment]
Same here, I don't understand it. And as for leveldb, it's supposed to be disabled for Android, so maybe cmake is overriding that. With that disabled I got an error with redis, with that disabled it was luajit next, with *that* disabled it's now showing
``In file included from jni/../jni/src/script/lua_api/l_areastore.cpp:26:
jni/src/util/areastore.h:34:11: fatal error: 'spatialindex/SpatialIndex.h' file not found
        #include <spatialindex/SpatialIndex.h>``. I ran "cmake ." in the minetest directory before, that's all I should need to do according to the other issue, right?

**Comment by stujones11** [comment]
I am pretty sure it is related to that issue and it would probably better to apply correct fix. Maybe I just got lucky, here's my `cmake_config.h` https://gist.github.com/stujones11/296f67ba2017e23dee87fea1b76e9d47

**Comment by Wayward1** [comment]
It seems you did, I replaced my cmake_config.h with yours and the build succeeded on the first try. Here's mine fwiw: https://gist.github.com/Wayward1/64a88d0b85a1da1e9aad7aa53e570843

**Comment by stujones11** [comment]
Yeah, thanks for the confirmation, I might submit a separate PR to fix that properly, though I thought one of the dev could have easily corrected this trivial bug without the need for another PR.

**Comment by Wayward1** [comment]
No problem :) And I can go ahead and submit a PR for that, I already have my git environment ready from another PR.

**Comment by stujones11** [comment]
@rubenwardy Is there any chance you could test this and confirm it fixes your issue? #6588

**Comment by nerzhul** [comment]
Merged as 22a891a925d68a096420d3383a2e31b58e3eb244 (little conflict with a previous commit i introduced

**Pull request closed without merging**